src/components/page-header/change-type-modal.tsx

```diff
@@ -0,0 +1,371 @@
+                    component: (
+                        <ThemeButton
+                            theme={newColor}
+                            onClick={() => {
+                                setIsNewThemeRequested(true);
+                                dispatch(openPaletteAppClip(theme));
+                            }}
+                        />
+                    ),
+                },
+            ],
+        },
+    ];
+
+    React.useEffect(() => {
+        if (isOpen) {
+            setIsStationTypeSwitch(false);
+            setIsLineStyleTypeSwitch(false);
+            setIsLinePathTypeSwitch(false);
+            setIsColorSwitch(false);
+            setThemeList([
+                {
+                    id: 'any',
+                    theme: [CityCode.Other, 'other', '#ffffff', MonoColour.black],
+                    value: t('header.settings.procedures.changeType.any'),
+                },
+                ...findThemes(graph.current, isSelect ? [...selected] : undefined).map(
+                    theme =>
+                        ({
+                            id: theme.toString(),
+                            theme: theme,
+                            value: theme[1] === 'other' ? theme[2] : theme[1],
+                        }) as ChangeTypeTheme
+                ),
+            ]);
+        }
+    }, [isOpen]);
+
+    const handleChange = () => {
+        if ((!filter || filter.includes('station')) && isStationTypeSwitch) {
+            changeStationsTypeInBatch(
+                graph.current,
+                currentStationType,
+                newStationType,
+                isSelect ? ([...selected].filter(node => node.startsWith('stn')) as StnId[]) : undefined
+            );
+        }
+        if ((!filter || filter.includes('line')) && isLineStyleTypeSwitch) {
+            changeLineStyleTypeInBatch(
+                graph.current,
+                currentLineStyleType,
+                newLineStyleType,
+                theme,
+                isSelect ? ([...selected].filter(edge => edge.startsWith('line')) as LineId[]) : undefined
+            );
+        }
+        if ((!filter || filter.includes('line')) && isLinePathTypeSwitch) {
+            changeLinePathTypeInBatch(
+                graph.current,
+                currentLinePathType,
+                newLinePathType,
+                isSelect ? ([...selected].filter(edge => edge.startsWith('line')) as LineId[]) : undefined
+            );
+        }
+        if (isColorSwitch) {
+            if (!filter || filter.includes('line'))
+                changeLinesColorInBatch(
+                    graph.current,
+                    currentColor,
+                    newColor,
+                    isSelect ? ([...selected].filter(edge => edge.startsWith('line')) as LineId[]) : undefined
+                );
+            if (!filter || filter.includes('station'))
+                changeNodesColorInBatch(
```

Comment:
Looks like we could only call this one time by passing stations and miscNodes with [...selected].filter.

Reply:
It may be complex. 👀 

```ts
if (!filter || filter.includes('misc-node') || filter.includes('station'))
                changeNodesColorInBatch(
                    graph.current,
                    currentColor,
                    newColor,
                    isSelect
                        ? !filter || filter.includes('station')
                            ? ([...selected].filter(node => node.startsWith('stn')) as StnId[])
                            : []
                        : undefined,
                    isSelect
                        ? !filter || filter.includes('misc-node')
                            ? ([...selected].filter(node => node.startsWith('misc_node')) as MiscNodeId[])
                            : []
                        : undefined
                );
```

Reply:
Yeah but you may compute stations via `const stations = filter?.includes('station') ? ([...selected].filter(node => node.startsWith('stn')) as StnId[]) : isSelect? graph.filterNodes(node => node.startsWith('stn')) : [];` and pass it in changeNodesColorInBatch if the `stations` could not be `undefined`.